### PR TITLE
[RLlib] Issue 24340: When multi-agent env run with only `default_policy`, env vs agent counters are incorrect.

### LINKED ISSUE #24340: [RLlib] wrong env step counting when train multi-agent with shared default policy
### What happened + What you expected to happen

When I train a multi-agent environment with a shared policy, I set:

```python
config = {
    ...
    'multiagent': {
        'policies': {
            'default_policy': PolicySpec(observation_space=None,
                                         action_space=None,
                                         config=None)
        }
    },
    ...
}
```

in my training config. Then the agents will use the same policy `default_policy`.

In the training result, the metric key `timesteps_total` means the number of sampled environment steps. And `agent_timesteps_total` means the sum of timesteps for all agents. I got:

```
agent_timesteps_total == timesteps_total
```

in my 8 agents environment.

Screen Shot

### Versions / Dependencies

ray[rllib] == 1.12.0

### Reproduction script

```python
import ray
from ray import tune
from ray.rllib.examples.env.multi_agent import MultiAgentCartPole
from ray.rllib.policy.policy import PolicySpec

config = {
    'env': MultiAgentCartPole,
    'env_config': {
        'num_agents': 4,
    },
    'multiagent': {
        'policies': {
            'default_policy': PolicySpec(observation_space=None,
                                         action_space=None,
                                         config=None)
        }
    },
    'framework': 'torch'
}

tune.run(
    'PPO',
    config=config,
    stop={'timesteps_total': 1E4}
)
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by sven1977**
Issue 24340: When multi-agent env run with only `default_policy`, env vs agent counters are incorrect.

## Why are these changes needed?
Issue #24340 

## Related issue number

Closes #24340
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by avnishn — APPROVED** [review]

**Comment by XuehaiPan** [comment]
Hi @sven1977, I created a PR #24342 for resolving this issue by checking `batch.env_step() == env_step` in:

```python
batch = MultiAgentBatch.wrap_as_needed(batch, env_steps=env_steps)
```

See https://github.com/ray-project/ray/pull/24342/files

We should update the tests `test_external_multi_agent_env_complete_episodes` and `test_external_multi_agent_env_truncate_episodes` in [`rllib/tests/test_external_multi_agent_env.py`](https://github.com/ray-project/ray/blob/master/rllib/tests/test_external_multi_agent_env.py)

File: rllib/evaluation/rollout_worker.py [hunk 1]
```diff
@@ -1035,20 +1035,20 @@ def compute_gradients(
         if log_once("compute_gradients"):
             logger.info("Compute gradients on:\n\n{}\n".format(summarize(samples)))
 
+        # Convert `samples` to a MultiAgentBatch, of not already.
```
Comment (gjoliver) [inline rllib/evaluation/rollout_worker.py:1038]: s/of/if not already.

**Review by gjoliver — APPROVED** [review]
one typo. thanks.

**Comment by sven1977** [comment]
Hey @XuehaiPan , cool, let me take a look at your PR.
I'm a little weary on changing the API for the sampler (returning only `MultiAgentBatch` instead of before a `Union[SampleBatch, MultiAgentBatch]`). Some tests are failing b/c of this.

...

**Pull request closed without merging**